File: win/MarkdownWin/MarkdownWin/FolderSearch.cs
//
//  FolderSearch.cs
//  MarkdownWin
//
//  Searches the open folder by file name and file contents. Mirrors FolderSearch.swift.
//
//  Deliberately does NOT go through the vault: md_vault_open() initialises git history and
//  records a baseline commit the first time a given path is opened, so routing search through
//  it would create a `.git` folder just from browsing — exactly what Workspace defers vault
//  access to avoid (see Workspace.GetOrOpenVault). This runs entirely off the UI thread.
//

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace MarkdownWin;

internal static class FolderSearch
{
    /// Stops one enormous file from stalling a search.
    private const long MaxFileBytes = 4 * 1024 * 1024;

    /// Matching lines kept per file.
    private const int MaxSnippetsPerFile = 3;

    /// Enough to fill the sidebar many times over; beyond this the query is too broad.
    private const int MaxHits = 200;

    private const int SnippetLimit = 160;

    /// <summary>Finds files whose name or contents contain <paramref name="query"/>.</summary>
    public static Task<List<SearchHit>> RunAsync(string root, string query, CancellationToken token) =>
        Task.Run(() => Run(root, query, token), token);

    private static List<SearchHit> Run(string root, string query, CancellationToken token)
    {
        string needle = query.Trim();
        if (needle.Length == 0)
        {
            return new List<SearchHit>();
        }

        var hits = new List<SearchHit>();

        foreach (string path in Candidates(root))
        {
            token.ThrowIfCancellationRequested();
            if (hits.Count >= MaxHits)
            {
                break;
            }

            long size;
            try
            {
                size = new FileInfo(path).Length;
            }
            catch (IOException)
            {
                continue;
            }

            bool nameMatched = Path.GetFileName(path).Contains(needle, StringComparison.CurrentCultureIgnoreCase);
            List<string> snippets = size <= MaxFileBytes ? MatchingLines(path, needle) : new List<string>();

            if (nameMatched || snippets.Count > 0)
            {
                hits.Add(new SearchHit(path, nameMatched, snippets));
            }
        }

        // Name matches first, then alphabetically, so results do not jump around.
        return hits
            .OrderByDescending(hit => hit.NameMatched)
            .ThenBy(hit => hit.Name, StringComparer.CurrentCultureIgnoreCase)
            .ToList();
    }

    /// <summary>Every Markdown file under <paramref name="root"/>, deepest paths included.</summary>
    private static IEnumerable<string> Candidates(string root)
    {
        var options = new EnumerationOptions
        {
            RecurseSubdirectories = true,
            IgnoreInaccessible = true,
            AttributesToSkip = FileAttributes.Hidden | FileAttributes.System,
        };

        IEnumerable<string> files;
        try
        {
            files = Directory.EnumerateFiles(root, "*", options);
        }
        catch (IOException)
        {
            return Enumerable.Empty<string>();
        }

        return files.Where(MarkdownFile.Matches);
    }

    private static List<string> MatchingLines(string path, string needle)
    {
        string[] lines;
        try
        {
            lines = File.ReadAllLines(path);
        }
        catch (IOException)
        {
            return new List<string>();
        }
        catch (UnauthorizedAccessException)
        {
            return new List<string>();
        }

        var snippets = new List<string>();
        foreach (string line in lines)
        {
            if (!line.Contains(needle, StringComparison.CurrentCultureIgnoreCase))
            {
                continue;
            }

            string trimmed = line.Trim();
            snippets.Add(trimmed.Length > SnippetLimit ? trimmed[..SnippetLimit] : trimmed);
            if (snippets.Count == MaxSnippetsPerFile)
            {
                break;
            }
        }

        return snippets;
    }
}

File: macos/Markdown/Markdown/FolderSearch.swift
//
//  FolderSearch.swift
//  Markdown
//
//  Searches the open folder by file name and file contents.
//
//  Runs off the main actor: a folder of notes can hold thousands of files, and reading
//  them must not stall typing in the search field.
//

import Foundation

/// One matching file, with the lines that matched.
nonisolated struct SearchHit: Identifiable, Sendable {
    let url: URL
    /// True when the file's name matched, which sorts it above content-only matches.
    let nameMatched: Bool
    /// Matching lines, trimmed for display.
    let snippets: [String]

    var id: URL { url }
    var name: String { url.lastPathComponent }
}

nonisolated enum FolderSearch {
    /// Stops one enormous file from stalling a search.
    private static let maxFileBytes = 4 << 20

    /// Matching lines kept per file.
    private static let maxSnippetsPerFile = 3

    /// Enough to fill the sidebar many times over; beyond this the query is too broad.
    private static let maxHits = 200

    private static let snippetLimit = 160

    /// Finds files whose name or contents contain `query`.
    ///
    /// Cooperatively cancellable: a new keystroke cancels the task running this.
    static func run(root: URL, query: String) async -> [SearchHit] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !needle.isEmpty else { return [] }

        var hits: [SearchHit] = []

        // Enumerated up front: a directory enumerator cannot be iterated from an async context.
        for url in candidates(under: root) {
            if Task.isCancelled { return [] }
            guard hits.count < maxHits else { break }

            let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey])
            guard values?.isRegularFile == true else { continue }

            let nameMatched = url.lastPathComponent.localizedCaseInsensitiveContains(needle)
            let size = values?.fileSize ?? 0
            let snippets = size <= maxFileBytes ? matchingLines(in: url, needle: needle) : []

            if nameMatched || !snippets.isEmpty {
                hits.append(SearchHit(url: url, nameMatched: nameMatched, snippets: snippets))
            }
        }

        // Name matches first, then alphabetically, so results do not jump around.
        return hits.sorted { lhs, rhs in
            lhs.nameMatched == rhs.nameMatched
                ? lhs.name.localizedStandardCompare(rhs.name) == .orderedAscending
                : lhs.nameMatched
        }
    }

    /// Every Markdown file under `root`, deepest paths included.
    private static func candidates(under root: URL) -> [URL] {
        let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey],
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        )
        guard let urls = enumerator?.allObjects as? [URL] else { return [] }
        return urls.filter(MarkdownFile.matches)
    }

    private static func matchingLines(in url: URL, needle: String) -> [String] {
        guard let contents = try? String(contentsOf: url, encoding: .utf8) else { return [] }

        var snippets: [String] = []
        for line in contents.split(separator: "\n", omittingEmptySubsequences: true) {
            guard line.localizedCaseInsensitiveContains(needle) else { continue }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            snippets.append(String(trimmed.prefix(snippetLimit)))
            if snippets.count == maxSnippetsPerFile { break }
        }
        return snippets
    }
}
